datafusion/common/src/tree_node.rs

```diff
@@ -145,40 +145,41 @@ pub trait TreeNode: Sized {
     /// Convenience utility for writing optimizer rules: Recursively apply the
     /// given function `f` to the tree in a bottom-up (post-order) fashion. When
     /// `f` does not apply to a given node, it is left unchanged.
-    fn transform<F: Fn(Self) -> Result<Transformed<Self>>>(
+    fn transform<F: FnMut(Self) -> Result<Transformed<Self>>>(
         self,
-        f: &F,
+        f: &mut F,
```

Comment:
I tried a bit locally to change this to be `f: F` (aka take an "owned" F) -- which would make it easier to call this API also works -- so  the callsites don't need to be changed to be `&mut` all over the place

However, I tried it and I got some strange errors when instantiating macros 🤔 

```
error: reached the recursion limit while instantiating `<expr::Expr as TreeNode>::transform_up::<&mut &mut &mut &mut ...>`
   --> /Users/andrewlamb/Software/arrow-datafusion/datafusion/common/src/tree_node.rs:184:50
    |
184 |         handle_transform_recursion_up!(self, |c| c.transform_up(&mut f), f)
    |                                                  ^^^^^^^^^^^^^^^^^^^^^^
    |
note: `transform_up` defined here
   --> /Users/andrewlamb/Software/arrow-datafusion/datafusion/common/src/tree_node.rs:180:5
    |
180 | /     fn transform_up<F: FnMut(Self) -> Result<Transformed<Self>>>(
181 | |         self,
182 | |         mut f: F,
183 | |     ) -> Result<Transformed<Self>> {
    | |__________________________________^
    = note: the full type name has been written to '/Users/andrewlamb/Software/arrow-datafusion/target/debug/deps/datafusion_expr-598aae957f8f2540.long-type.txt'


```

Reply:
Yes, I did try this already and run into the same issue...
But now I think it is doable with some smart refactor. Let me update this PR today or tomorrow and then we can decide if this change make sense or not

Reply:
https://github.com/apache/arrow-datafusion/pull/10126/commits/d781ca714fe1ec6be25aa20f6bb15fdfc1419641 changes `TreeNode::apply()`, `TreeNode::transform()`, `TreeNode::transform_down()`, `TreeNode::transform_up()`, `TreeNode::transform_down_up()` and their `LogicalPlan::...with_subqueries()` variants to use owned closures.

Reply:
If we like this change I can update the PR description.